File: app/src/serverApi/IMetric.ts
import { MetricType } from "./MetricType";
import { IMetricAttributes } from "./IMetricAttributes";
import { IUserPermissions } from "./IUserPermissions";
import { IMetricThresholds } from "./IMetricThresholds";
import { IMetricCustomProps } from "./IMetricCustomProps";

export interface IMetric {
  id?: string;
  attributes?: IMetricAttributes;
  name?: string;
  description?: string;
  notes?: string;
  type: MetricType;
  editedOn?: string;
  thresholds?: IMetricThresholds;
  permissions?: IUserPermissions;
  customProps?: IMetricCustomProps;
}

File: api/Metrix.Core/Source/Domain/Metrics/IMetric.cs
using Metrix.Core.Domain.Permissions;

namespace Metrix.Core.Domain.Metrics;

public interface IMetric : IUserScoped, IPermissionHolder
{
  string? Id { get; set; }

  string Name { get; set; }

  string? Description { get; set; }

  string? Notes { get; set; }

  MetricType Type { get; }

  Dictionary<string, MetricAttribute> Attributes { get; set; }

  Dictionary<string, Dictionary<string, double>> Thresholds { get; set; }

  DateTime? EditedOn { get; set; }

  // basic idea here is to enable clients to store whatever they want while
  // the server is completely unaware of what's going on here. i.e. server
  // does not know, what he's actually storing. this way there's also no
  // server side serialization required, etc.
  Dictionary<string, string> CustomProps { get; set; }
}
